address_to_census_tract.py

```diff
@@ -7,81 +7,194 @@
-    ]
+    return standard_address + ', ' + response['last_line']
 
+def response_to_latlng(response: object) -> list:
+    """
+    Returns latitude and longitude coordinates from the given HTTP response
+    object.
+    """
+    return [response['metadata']['latitude'], response['metadata']['longitude']]
 
-def latlng_to_polygon(latlng, polygons):
+def latlng_to_polygon(latlng):
     """
-    Find the first polygon in *polygons* which contains the *latlng* and return
-    the polygon, else None.
+    Find the first polygon in *polygons* (loaded from file) which contains the 
+    *latlng* and return the polygon, else None.
     """
+    polygons = load_geojson("data/geojsons/Washington_2016.geojson")
 
     # Ye olde lat/lng vs. lng/lat schism rears its head.
     lat, lng = latlng
     point = Point(lng, lat)
 
     for polygon in polygons:
         if polygon["shape"].contains(point):
-            return polygon
+            return polygon.get("properties", {}).get("GEOID")
 
+    print(f"failed to find tract for {latlng}", file = stderr)
     return None
 
+def reformat_address(address: str) -> str: 
+    """
+    Removes a stray unit/suite number at the front of a given address. Also
+    removes unit/suite numbers followed by pound sign (#). These unit/suite 
+    numbers often cause errors in geocoding.  
+
+    Assumptions:
+    1. A street address contains at least one letter, [a-z] (case insensitive).
+    2. A street address comes before a city, state, and zip code in an address.
+    3. An address won't have more than one comma-separated string ahead of a 
+    true street address.
+
+    The following method should work even if the given address has no commas
+    separating the address components. 
+    """
+    address = address.lower()
+    address_pieces = address.split(',')
+
+    # Check first piece. If it does not look like a street, drop it.
+    street = address_pieces[0]
+    contains_letter = bool(re.search('[a-z]', street))
+    if not contains_letter and len(address_pieces) > 1:
+        address_pieces.remove(street)
+    
+    # Check first piece (should be address). If it has a '#', drop the
+    # entire unit or suite number, etc.
+    regexp = re.search('#.*(\s?)', street)
+    contains_unit = bool(regexp)
+    if contains_unit: 
+        unit_span = regexp.span()
+        address_before_unit = street[0:unit_span[0]]
+        address_after_unit = street[unit_span[1]:]
+
+        new_street = address_before_unit + address_after_unit
+        address_pieces[0] = new_street
+
+    return ''.join(address_pieces)
+
+def build_request_url(address: str) -> str:
+    """
+    Given a dict containing an address, builds a request URL for the
+    SmartyStreets geocoding API
+    """
+    NUM_CANDIDATES = 1
+
+    url = 'https://us-street.api.smartystreets.com/street-address?'
+    url += f'auth-id={AUTH_ID}&auth-token={AUTH_TOKEN}&street='
+    url += f'{address}&candidates={NUM_CANDIDATES}'
```

Comment:
The interpolated values aren't URL-encoded, which means they'll break if the address (or id/token) contains a reserved character, such as in `4th & Walnut`. Note that `requests` supports ways to pass a dict to generate query or body params, but also that this function won't be necessary if we use the SmartyStreets SDK.

Reply:
Resolving in favor of using SmartyStreets SDK.